### PR TITLE
cryptsetup: detached header verification, refactor

### PR DESCRIPTION
**Pull request by katexochen**
With this change, we start verifying the cryptsetup header and use it in detached mode afterwards. 

This is a breaking change for cryptsetup volumes that were set up with earlier versions of Contrast or manually with different cryptsetup parameters.

### CODE REVIEW

**Review by charludo — COMMENTED** [review]
I will have another look in the morning, but so far everything seems good. Two "features" I think might be nice-to-have:

- for one-off mounts, it would be nice to not have to specify a `headerPath`. A dummy one works as well, of course, but is maybe not as clean.
- closing the device: would it maybe make sense to call `luksClose` on error?

File: internal/cryptsetup/cryptsetup.go [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+	if digest.Digest == "" {
+		return fmt.Errorf("expected digest to be non-empty")
+	}
+	if len(header.Tokens) != 0 {
+		return fmt.Errorf("expected no tokens, got %d", len(header.Tokens))
+	}
+	return nil
+}
+
+type cryptsetupMetadata struct {
```
Comment (burgerdev) [inline internal/cryptsetup/cryptsetup.go:258]: Is there some source for this structure that we could cite here, or do we only know it empirically?

File: internal/cryptsetup/cryptsetup.go [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+		Segments   []string `json:"segments"`
+		Hash       string   `json:"hash"`
+		Iterations int      `json:"iterations"`
+		Salt       string   `json:"salt"`
+		Digest     string   `json:"digest"`
+	} `json:"digests"`
+	Config struct {
+		JSONSize     string `json:"json_size"`
+		KeyslotsSize string `json:"keyslots_size"`
+	}
```
Comment (burgerdev) [inline internal/cryptsetup/cryptsetup.go:304]: `Config` does not appear to be checked, I suppose because it's dynamic of sorts?

File: internal/cryptsetup/cryptsetup.go [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+	if errors.As(err, &exitErr) {
+		return "", fmt.Errorf("executing '%s': %w, stderr: %s", cmd.String(), err, exitErr.Stderr)
+	} else if err != nil {
+		return "", fmt.Errorf("executing '%s': %w", cmd.String(), err)
+	}
+	return string(output), nil
+}
+
+func ensureBase(p string) error {
+	if err := os.MkdirAll(filepath.Dir(p), 0o700); err != nil {
```
Comment (burgerdev) [inline internal/cryptsetup/cryptsetup.go]: I think inlining this would be more readable.

File: internal/cryptsetup/cryptsetup.go [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+	}
+	args := []string{
+		"luksFormat",
+		"--type=luks2",                          // Use LUKS2 header format.
+		"--cipher=aes-xts-plain64",              // Use AES-XTS cipher.
+		"--pbkdf=argon2id",                      // Use Argon2id as the key derivation function.
+		"--pbkdf-memory=10240",                  // Memory usage for Argon2i, limit to 10 MiB so it won't fail in low-memory pods.
+		"--batch-mode",                          // Suppresses all confirmation questions.
+		fmt.Sprintf("--key-file=%s", d.keyPath), // Path to the key file.
+		d.devicePath,
```
Comment (burgerdev) [inline internal/cryptsetup/cryptsetup.go:69]: Since the parameters are fixed, we should be able to tell whether this is compatible or not with prior cryptsetup volumes (referring to the "might be breaking" in the PR description).

File: internal/cryptsetup/cryptsetup_test.go [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+      "digest":"gzHnbLzZpMe0tMMeP52UgBIjpQDgyP6J6y4St6sX+0M="
+    }
+  },
+  "config":{
+    "json_size":"12288",
+    "keyslots_size":"16744448"
+  }
+}`
+
+func TestVerifyHeader(t *testing.T) {
```
Comment (burgerdev) [inline internal/cryptsetup/cryptsetup_test.go:73]: We don't need to test every single header field, but a handful of negative tests would be reassuring.

File: initializer/main.go [hunk 6]
```diff
@@ -184,27 +186,32 @@ func run(cmd *cobra.Command, _ []string) (retErr error) {
...
 
-	flags := &cryptsetupFlags{
-		devicePath:       cryptsetupDevicePath,
-		volumeMountPoint: "/state",
+	if cryptsetupDevicePath != "" {
+		// The device mount is created by the initializer and consumed by an application container.
+		// This is only possible through mount propagation, from the initializer mount namespace to
+		// the other mount namespaces of the pod. If the initializer exists, its mount namespace is
+		// removed, which also removes the propagated mount. So we must idle instead of exiting.
+		log.Info("Ideling to keep propagated cryptsetup device mounts alive")
```
Comment (burgerdev) [inline initializer/main.go]: Idling.

File: initializer/main.go [hunk 7]
```diff
@@ -184,27 +186,32 @@ func run(cmd *cobra.Command, _ []string) (retErr error) {
...
+	log.Info("Initializer done")
 
-	flags := &cryptsetupFlags{
-		devicePath:       cryptsetupDevicePath,
-		volumeMountPoint: "/state",
+	if cryptsetupDevicePath != "" {
+		// The device mount is created by the initializer and consumed by an application container.
+		// This is only possible through mount propagation, from the initializer mount namespace to
+		// the other mount namespaces of the pod. If the initializer exists, its mount namespace is
+		// removed, which also removes the propagated mount. So we must idle instead of exiting.
```
Comment (burgerdev) [inline initializer/main.go]: This comment slightly misses the underlying issue, which afaiu is that we're mounting on the tmpfs mountpoint itself. If you change the setup such that `/mnt` holds the tmpfs and `/mnt/state` the LUKS device mount, the initializer does not need to keep running. To be honest, I was surprised that the current approach works at all.

**Review by burgerdev — APPROVED** [review]
Very nice package, thanks a lot!

File: internal/cryptsetup/cryptsetup.go (hunk 4 above)
Comment (katexochen) [reply internal/cryptsetup/cryptsetup.go:69]: It will break with the follow up PR (integrity).

File: internal/cryptsetup/cryptsetup_test.go (hunk 5 above)
Comment (katexochen) [reply internal/cryptsetup/cryptsetup_test.go:73]: Will do in the follow up.

File: internal/cryptsetup/cryptsetup.go (hunk 2 above)
Comment (katexochen) [reply internal/cryptsetup/cryptsetup.go:304]: I think it is. Not sure why you should verify JSONSize (might make sense if you are writing C?).
In general, the current code doesn't verify size/offset values beside `key_size`.

**Comment by burgerdev** [comment]
> > for one-off mounts, it would be nice to not have to specify a headerPath. A dummy one works as well, of course, but is maybe not as clean.
> 
> Thinking about it, this shouldn't be exposed at all.

There are requirements for the detached header placement (i.e., in a secure, writable location). Only the caller knows what these are, although granted we control who's calling it so a sane default would work, too.

**Comment by katexochen** [comment]
> There are requirements for the detached header placement (i.e., in a secure, writable location). Only the caller knows what these are, although granted we control who's calling it so a sane default would work, too.

I moved the location to /dev/shm to guarantee the header is placed in memory. That location should also be writable.

File: initializer/main.go (hunk 7 above)
Comment (burgerdev) [reply initializer/main.go]: Closer to the truth:

> The device mount is created by the initializer and shared with the application container through a common emptyDir. We want to avoid mounting on a sub-path of the emptyDir to make configuration for the application container easier, so we mount the encrypted volume on top of the emptyDir mount point. The Kata agent does not know about this and will unmount that mount point when the initializer exits, which affects the application container, too. We avoid this situation by keeping the initializer running indefinitely.

File: internal/cryptsetup/cryptsetup.go [hunk 8]
```diff
@@ -0,0 +1,287 @@
...
+	headerPath string
+	keyPath    string
+}
+
+// NewDevice creates a new Device instance. It doesn't interact with the device itself.
+func NewDevice(devicePath, keyPath string) (*Device, error) {
+	var randSuffix [4]byte
+	_, err := rand.Read(randSuffix[:])
+	if err != nil {
+		return nil, fmt.Errorf("generating random suffix for key file: %w", err)
```
Comment (charludo) [inline internal/cryptsetup/cryptsetup.go]:
```suggestion
		return nil, fmt.Errorf("generating random suffix for header file: %w", err)
```

**Review by burgerdev — APPROVED** [review]
